jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java

```diff
@@ -0,0 +1,6132 @@
+	 */
+	public int arrayBaseOffset(Class c) {
+		Objects.requireNonNull(c);
+		return arrayBaseOffset0(c);
+	}
+
+	/**
+	 * Return index size of array in bytes.
+	 * 
+	 * @param c class array
+	 * @return returns index size of array in bytes
+	 * 
+	 * @throws NullPointerException if class is null
+	 * @throws IllegalArgumentException if class is not an array
+	 */
+	public int arrayIndexScale(Class c) {
+		Objects.requireNonNull(c);
+		return arrayIndexScale0(c);
+	}
+
+	/**
+	 * @return size of address on machine in use
+	 */
+	public int addressSize() {
+		return ADDRESS_SIZE;
+	}
+
+	/**
+	 * Creates a class out of a given array of bytes with a ProtectionDomain.
+	 * 
+	 * @param name binary name of the class, null if the name is not known
+	 * @param b a byte array of the class data. The bytes should have the format of a 
+	 * 			valid class file as defined by The JVM Spec
+	 * @param offset offset of the start of the class data in b
+	 * @param bLength length of the class data
+	 * @param cl ClassLoader used to load the class being built. If null, the default 
+	 * system ClassLoader will be used
+	 * @param pd ProtectionDomain for new class
+	 * @return class created from the byte array and ProtectionDomain parameters
+	 * 
+	 * @throws NullPointerException if b array of data is null
+	 * @throws ArrayIndexOutOfBoundsException if bLength is negative
+	 * @throws IndexOutOfBoundsException if offset + bLength is greater than the 
+	 * length of b
+	 */
+	public Class<?> defineClass(String name, byte[] b, int offset, int bLength, ClassLoader cl, ProtectionDomain pd) {
+		Objects.requireNonNull(b);
+
+		if (bLength < 0) {
+			throw new ArrayIndexOutOfBoundsException();
+		}
+
+		return defineClass0(name, b, offset, bLength, cl, pd);
+	}
+
+	/**
+	 * Define a class without making it known to the class loader.
+	 * 
+	 * @param hostingClass the context for class loader + linkage, and 
+	 * access control + protection domain
+	 * @param bytecodes class file bytes
+	 * @param constPatches entries that are not "null" are replacements 
+	 * for the corresponding const pool entries in the 'bytecodes' data
+	 * @return class created from bytecodes and constPatches
+	 * 
+	 * @throws NullPointerException if hostingClass or bytecodes is null
+	 * @throws IllegalArgumentException if hostingClass is an array or primitive
+	 */
+	public Class<?> defineAnonymousClass(Class<?> hostingClass, byte[] bytecodes, Object[] constPatches) {
+		Objects.requireNonNull(hostingClass);
+		Objects.requireNonNull(bytecodes);
+		
+		if (hostingClass.isArray() || hostingClass.isPrimitive()) {
+			throw invalidInput();
+		}
+
+		return defineAnonymousClass0(hostingClass, bytecodes, constPatches);
+	}
+
+	/**
+	 * Allocate new array of same type as class parameter and 
+	 * length of int parameter.
+	 * 
+	 * @param c class of same type as desired array
+	 * @param length desired length of array
+	 * @return allocated array of desired length and type, or null 
```

Comment:
"or null is type os not a primitive" doesn't match the actual behavior

Reply:
"or null if class type is not a primitive wrapper" ?

Reply:
Can it ever return null? It throws IllegalArgumentException.

Reply:
Oh right. No it cannot.